### ISSUE TITLE
First Training - Estimate time is TOO SLOW with RTX 5090 - I'm doing it wrong... HELP 🙏

### ISSUE THREAD

**Issue by AlonDan** [kohya-ss/musubi-tuner#269]
Hello All,
After a LONG struggle trying to install Musubi-Tuner on Windows 11 + RTX-5090 I finally did it! 👍
I'm planning to share step-by-step guide on how I installed it so it can (hopefully) help anyone who struggle like me.
I will also add the BATCH files I created (one-click to: cache latents videos + txt, one click to start train, one click to convert to ComfyUI LoRA) 
Stay tuned...

---

What I did first, I just made some BATCH files, and .TOML  to make my life easier.

### My First TEST:
The goal is to test if my Musubi-Tuner works, after the rough installation.

On my VERY first TEST, I placed 1 video.mp4 of few seconds 512x512, 16 fps with a .TXT caption, on my dataset videos folder
I created a .TOML file basically with: "EPOCHS=1" 
I Cached the latents of both VIDEO + TEXT. (my batch file do both in one click)
Then I TRAINED the tiny thing, it took about 1 minutes (my batch file also tells me how long it took)

So far so good, it was created, means I can TRAIN my actual first Wan 2.1 i2v LoRA.

---

### My first Wan 2.1 (i2v) LoRA Training:
The goal is to train focus on MOTION:

**NOTE:** 
This is the very first time I'm trying to train Wan 2.1 (i2v) so I'm still reading the DOCS and other forums / discord to try understand about the NUMBERS, at the moment I went with something that **CoPilot** suggested me based on my SPECS and goal, which is focus on MOTIONS.

I bet I'm doing HUGE mistakes here already from numbers in the settings, toml, dataset, etc.. but this is why I'm asking here how to make it faster yet not LOSE quality or ruin the training, I'm here to learn from YOUR experience, especially anyone who already trained (i2v) or at least know a thing about what should I do, all suggestions will be very appreciated! 🙏

---
### DATASET:

8 x Text files (.txt) for caption of each video
8 x Video files (.mp4) 
512x512 resolution
16 fps
48 frames each video

---
### TOML:

[[datasets]]
video_directory = "dataset/videos"
cache_directory = "dataset/videos/cache_latents"
caption_extension = ".txt"
enable_bucket = true
resolution = [512, 512]
target_frames = [48]
frame_extraction = "full"

### TRAINING Settings (batch file)
It's from my batch file so it looks different, I cut out stuff to keep it as clean as possible, sorry about that...

set EPOCHS=30
set LORA_NAME=My_First_LoRA_wan_i2v_14b
   
    --task i2v-14B ^
    --dit ./models/wan/wan2.1_i2v_720p_14B_bf16.safetensors ^
    --dataset_config "./dataset/Kneel_Bounce_v1.toml" ^
    --mixed_precision bf16 ^
    --optimizer_type adamw8bit ^
    --learning_rate 2e-4 ^
    --gradient_checkpointing ^
    --max_data_loader_n_workers 2 ^
    --persistent_data_loader_workers ^
    --network_module networks.lora_wan ^
    --network_dim 32 ^
    --timestep_sampling shift ^
    --discrete_flow_shift 3.0 ^
    --max_train_epochs %EPOCHS% ^
    --save_every_n_epochs 10 ^
    --seed 42 ^
    --output_dir "./Output_LoRAs/" ^
    --output_name %LORA_NAME% ^
    --flash_attn ^
    --network_args "loraplus_lr_ratio=4" ^
    --fp8_base

---

At the moment it's already running for about 40 minutes and this is what it shows...
Basically it's on the 6th epoch out of 240... you got the idea!
I'm doing something VERY wrong and you probably already notice it based on the settings I post above.

![Image](https://github.com/user-attachments/assets/dcfb4585-cb42-4294-a838-a110a4ba4ad5)

---

As you can see based on all the information, I need YOUR help here, so I can train the SAME dataset, but with time that make sense,
It's important that I won't lose quality of the output LoRA, even that it's MOTION focused I want to test it in ComfyUI and see that it works as expected so I can LEARN how to improve.

But first... how do I make the training TIME to something that take few hours, and not... YEARS with my specs at least.

Thanks ahead to anyone who can help, please feel free to share edited .TOML or SETTINGS, etc.. and if possible please EXPLAIN since I'm a noob and don't understand every single number what it ACTUALLY means, same for every setting and model related choices.
The goal is to train: Wan Image to Video (i2v) LoRA that will focus on MOTION from the dataset.

---

### My PC specs:
Intel Core Ultra 9 285K
Nvidia RTX 5090
96GB Ram
Windows 11

**Comment by maybleMyers** [comment]
I would bet you are not using enough block swap. If you go into shared vram in windows it will drastically slow things down.

**Comment by Sarania** [comment]
^ Agreed. Doesn't 5090 have 32GB though? But yeah if you go into shared in Windows it will KILL your perf. Use --blocks_to_swap maybe start at 20 but yeah aim to have around ~1GB free so you have a lil buffer to avoid OOM. Look in your task manager to check VRAM / shared VRAM as well!

Side tips: You are using the bf16 model with bf16 precision, but we've discovered that Wan performs better in fp16 ( sauce: https://comfyui.org/en/wan-2-1-hunyuan-video-generation-updates ) so you might consider using the fp16 model with fp16. Even though you are using fp8_base in the end, it still matters! Secondly, I HIGHLY recommend using `--fp8_scaled`. It does a scaled quantization instead of a naive one and there's no real downside it's just closer to the unquantized model! Lastly IMO that LR is WAY high for Wan. I personally use 2e-5 with a lora plus of 4, without lora plus you might do 5e-5 - 8e-5

**Comment by AlonDan** [comment]
Thank you for your tips, I will try to play with the numbers and models such as fp16 as you mentioned. ❤️

Before I saw your helpful replies, I used CoPilot (Think Deeper mode) since it helps me in most cases so far but I'm still learning and experimental just began since training takes time... 

---

So far I managed to train my very first test with the same dataset of 8 very short videos, 512x512 etc.

I did my first train and it seems like it's working nicely so far, quality is something I'm checking as well.
I trained on:  `wan2.1_i2v_720p_14B_bf16.safetensors` as you already mentioned I should try **fp16** next time ✅
Consider I'm now generating on 512x512 (same as the datasource trained) but that's not the reason, it's because it's easier for me to crop 1:1 ratio and do different length tests, of course I will also need to test 720p res and ratio.
But so far,
The motion is pretty much what I expect, so far so good for my very first training. I tried 20 and 30 steps, even on 20 it's nice!

Just sharing the time it took and such on the new CHANGES I did with CoPilot (totally different from the above)

Since I'm a designer in spirit but not a programmer, 
It's something I added to the BATCH (one-click training) so it's nice to see the results:
![Image](https://github.com/user-attachments/assets/31c445ff-bb6e-4587-ade5-af154090cf21)

I basically created 3 batch files:
1 - Create Cache Latents for Video and Text (one click, done!)
2 - TRAIN the LoRA - (this one is instead of .json so it needs to EDIT the settings inside per NEW training)
3 - Convert to ComfyUI LoRA - (just drag and drop your files, it will do the magic)

I created MORE batch files for other tasks for example: 
Launch TensorBoard in 1 click and open in the Browser but like I said it doesn't "look" at the training so I'm not using it for now.
And more batch files for QOL because it's easy for me.

Basically, I need to edit 3 files every NEW train from now on:
- the TRAIN batch file 
- The dataset .TOML 
- Sample Prompts text file
And supply 1 sample image reference for Sample Reference.

I still didn't understand what it does because it created 473 images in my case (based on the settings).
I guess it's not for ME to check but to HELP the training itself, if I'm not wrong... that's what CoPilot explained at least.


As you can see I used **50 EPOCHS** this time since my videos are very short I wanted to try.

---

Here are the settings I used (it's from my BATCH file so I cut only the part with the settings, sorry for the mess)

EPOCHS=50
LORA_NAME=My_First_LoRA_wan_i2v_14b

call venv\Scripts\accelerate.exe launch --num_cpu_threads_per_process 1 ^
    wan_train_network.py ^
    --task i2v-14B ^
    --dit ./models/wan/wan2.1_i2v_720p_14B_bf16.safetensors ^
    --clip "./models/wan/models_clip_open-clip-xlm-roberta-large-vit-huge-14.pth" ^
	--t5 "./models/wan/models_t5_umt5-xxl-enc-bf16.pth" ^
    --vae "./models/wan/wan_2.1_vae.safetensors" ^
    --dataset_config "./dataset/Kneel_Bounce_v1.toml" ^
    --mixed_precision bf16 ^
    --optimizer_type adamw8bit ^
    --learning_rate 2e-4 ^
    --gradient_checkpointing ^
    --max_data_loader_n_workers 16 ^
    --persistent_data_loader_workers ^
    --network_module networks.lora_wan ^
    --network_dim 32 ^
    --timestep_sampling shift ^
    --discrete_flow_shift 3.0 ^
    --max_train_epochs %EPOCHS% ^
    --max_train_steps 2000 ^
    --save_every_n_epochs 10 ^
    --save_every_n_steps 100 ^
    --sample_prompts "./dataset/___SAMPLE_Prompts.txt" ^
    --sample_every_n_steps 50 ^
    --sample_at_first ^
    --seed 42 ^
    --output_dir "./Output_LoRAs/" ^
    --output_name %LORA_NAME% ^
    --flash_attn ^
    --network_args "loraplus_lr_ratio=4" ^
    --fp8_base ^
    --fp8_scaled ^
    --sdpa ^
    --logging_dir "./logs/"


---

The problem:
It works fine, but I'm not sure if "FINE" is the best result I can get, or should I try more EPOCHS for focus on MOTIONS training?
Sure, if it works don't fix it... 👍 but I'm still testing, I think it works I just can't tell if it's because 50 epochs were enough, or because it's just a motion so it's "easier" to train? I have no clue... just sharing my mind while I'm going to experiment more.

---

I tried to use TensorBoard so I could SEE or FOLLOW what's goin on during the training, it seems like it's installed fine and it appears on the browser but... it's never refreshed, even if I change to other tabs. 
I tried to launch it before and while training tests, nothing helped.
I also tried to fix with the help of: CoPilot and ChatGPT try makingit work with the training... I failed for over and over so I gave up on **TensorBoard** for now, I guess it's not that simple with Musubi-Tuner at the moment

I believe I should use TensorBoard (not that I know how to use it's functions, but it should look clean enough for visual people such as myself). 

---
CoPilot explained to me that bf16 will give much more accurate results especially for motion so... I kinda listen to it, and so far it didn't disappoint but if I get it right **fp16** with **--fp8_scaled** may do the SAME but quicker and even better? (If I got it right from the other discussion) 🤔


With all the respect to CoPilot or other chatbots, YOUR actual experiment is the best to learn from!
I'm just using it because it helps explain some of the things I'm not sure about or confused, sure I have SO MANY MORE questions about training and experimenting in the next LoRA, motion related and non-motion as well... 

Please feel free to share YOUR tips & tricks based on your experience, 
I'm learning a lot from the community thanks everyone 🙏

**Comment by Sarania** [comment]
CoPilot told you that about bf16 because that is USUALLY the case with ML models - bf16 has a bigger range, fp16 has more precision. Usually models care more about their outliers and range than raw precision so bf16 is better. Wan is different for some reason(it also doesn't like fp8_fast/mm_scaled which uses a combo of fp8_e4m3fn and fp8_e5m2 for likely the same reason, with e5m2 being too much precision loss.) As far as what it will get you, just slightly better quality closer to the full fp32 model. Both for inference and training. 

As far as your results, training is as much an art as a science so there is definitely some experimenting to be done. But ultimately if you are happy with what you made that's what should tell if you it's good enough or not!

**Comment by sdbds** [comment]
> CoPilot told you that about bf16 because that is USUALLY the case with ML models - bf16 has a bigger range, fp16 has more precision. Usually models care more about their outliers and range than raw precision so bf16 is better. Wan is different for some reason(it also doesn't like fp8_fast/mm_scaled which uses a combo of fp8_e4m3fn and fp8_e5m2 for likely the same reason, with e5m2 being too much precision loss.) As far as what it will get you, just slightly better quality closer to the full fp32 model. Both for inference and training.
> 
> As far as your results, training is as much an art as a science so there is definitely some experimenting to be done. But ultimately if you are happy with what you made that's what should tell if you it's good enough or not!

Actually FP16 is much better than BF16... but due to various issues, there's a bug in T5 where it produces NAN in fp16, so now all our T5 models must use BF16...
